funsor/optimizer.py

```diff
@@ -1,197 +1,100 @@
 
+    for i, v in enumerate(terms):
 
-@dispatched_interpretation
-def distribute(cls, *args):
-    result = distribute.dispatch(cls, *args)
-    if result is None:
-        result = lazy(cls, *args)
-    return result
+        if not isinstance(v, Contraction):
+            continue
 
+        if v.red_op is nullop and (v.bin_op, bin_op) in DISTRIBUTIVE_OPS:
+            # a * e * (b + c + d) -> (a * e * b) + (a * e * c) + (a * e * d)
+            new_terms = tuple(
+                Contraction(v.red_op, bin_op, v.reduced_vars, *(terms[:i] + (vt,) + terms[i+1:]))
+                for vt in v.terms)
+            return Contraction(red_op, v.bin_op, reduced_vars, *new_terms)
 
-@distribute.register(Finitary, AssociativeOp, tuple)
-def distribute_finitary(op, operands):
-    # TODO raise an error or warning on name collision
-    if len(operands) == 1:
-        return operands[0]
-
-    reduce_op = None
-    reduce_terms, remaining_terms, reduced_vars = [], [], frozenset()
-    for term in operands:
-        # term is not reduce -> do nothing
-        # term is reduce but does not distribute -> do nothing
-        # term is reduce and distributes -> put into reduce_terms
-        if isinstance(term, Reduce):
-            if not reduce_op or not (reduce_op, op) in DISTRIBUTIVE_OPS:
-                reduce_op = term.op
-            if term.op == reduce_op and (reduce_op, op) in DISTRIBUTIVE_OPS:
-                reduce_terms.append(term)
-                reduced_vars = reduced_vars | term.reduced_vars
-            else:
-                remaining_terms.append(term)
-        else:
-            remaining_terms.append(term)
-
-    if len(reduce_terms) > 1:
-        new_finitary_term = Finitary(op, tuple(term.arg for term in reduce_terms))
-        remaining_terms.append(Reduce(reduce_op, new_finitary_term, reduced_vars))
-        return Finitary(op, tuple(remaining_terms))
+        if red_op in (v.red_op, nullop) and (v.red_op, bin_op) in DISTRIBUTIVE_OPS:
+            new_terms = terms[:i] + (Contraction(v.red_op, v.bin_op, frozenset(), *v.terms),) + terms[i+1:]
+            return Contraction(v.red_op, bin_op, v.reduced_vars, *new_terms).reduce(red_op, reduced_vars)
+
+        if v.red_op in (red_op, nullop) and bin_op in (v.bin_op, nullop):
+            red_op = v.red_op if red_op is nullop else red_op
+            bin_op = v.bin_op if bin_op is nullop else bin_op
+            new_terms = terms[:i] + v.terms + terms[i+1:]
+            return Contraction(red_op, bin_op, reduced_vars | v.reduced_vars, *new_terms)
 
     return None
 
 
-@dispatched_interpretation
+unfold.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Variadic[Funsor])(
+    lambda r, b, v, *ts: unfold(Contraction, r, b, v, tuple(ts)))
+
+
+@interpreter.dispatched_interpretation
 def optimize(cls, *args):
     result = optimize.dispatch(cls, *args)
     if result is None:
-        result = lazy(cls, *args)
+        result = eager(cls, *args)
```

Comment:
Does this mean we can eliminate lots of `reinterpret` calls in examples?

Reply:
I think so, unless you want the more aggressive rewrite performed by `apply_optimizer`